swift/extractor/visitors/VisitorBase.h

```diff
@@ -21,14 +21,34 @@ class VisitorBase {
 
 // we want to override the default swift visitor behaviour of chaining calls to immediate
 // superclasses by default and instead provide our own TBD default (using the exact type)
-#define DEFAULT(KIND, CLASS, PARENT) \
-  void visit##CLASS##KIND(swift::CLASS##KIND* e) { dispatcher_.emitUnknown(e); }
+// if we the implementation class has translate##CLASS##KIND (that uses generated C++ classes), we
+// want to use that. We detect that by comparing the member function pointer to the one for a
+// private undefined member function of the same name and signature in {Ast,Type}VisitorBase, which
+// will be shadowed (and thus different) in case the implementation class has it.
+#define DEFAULT(KIND, CLASS, PARENT)                                              \
+ public:                                                                          \
+  void visit##CLASS##KIND(swift::CLASS##KIND* e) {                                \
```

Comment:
Even with the comment (which is very helpful!), this makes me a bit uneasy: can we actually compare pointers to members of different types? even if they have different access? will they definitely never compare equal, or does the base one not being defined give the implementation some leeway that we're not expecting?

I can live with it, but would suggest thinking about alternatives:
1. Instead of `translate##CLASS##KIND` taking an output parameter (see also my comment on the declaration), can it return the TRAP class instead? If yes, the base implementation can return `void` (or some other sentinel - even a runtime sentinel like `nullptr` or `nullopt`), and we can check for that here. That also has the potentially useful property of subclasses being able to return void when they want to do something but not emit TRAP.
2. Can we add a base implementation that does the "right" thing (i.e. populates a TrapClassOf that causes `emit` to act as `emitUnknown`)?
3. If we keep the check, it might be more obviously correct with SFINAE or `std::is_invocable`?